### PR TITLE
feat(ts): auto generate tsconfig.json

### PR DESCRIPTION
**Pull request by kevinmarrec**
## Types of changes

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description
This PR implements the autogeneration of a default `tsconfig.json` (if no one exists):
```json
{
  "extends": "@nuxt/typescript",
  "compilerOptions": {
    "baseUrl": "."
  }
}
```
It extends a base `tsconfig.json` defined in a new `@nuxt/typescript` package, shipped by `nuxt-ts` distribution.

`baseUrl` is mandatory cause of a relativePaths thing.

## Todo
- [x] Define a default/base `tsconfig.json` which all projects can now extend from
- [x] Ship `@types/node` dependency and all TypeScript dependencies under a new `@nuxt/typescript` package
- [x] Reference `@types/node` & `@nuxt/vue-app` in base `tsconfig.json`
- [x] Update examples and fixtures `tsconfig.json` files
- [x] Find a way to make base `tsconfig.json` reference edge versions of types when publishing edge version
- [x] Looks for the right `tsconfig.json` path when using `nuxt-ts path/to/project`
- [x] Implement a prompt which ask you confirmation for generating the `tsconfig.json` if it doesn't exist
- [x] Add a tests which check the file creation/content and that `ts-node` is registered

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [x] I have added tests to cover my changes (if not applicable, please state why)
- [x] All new and existing tests are passing.

### CODE REVIEW

**Comment by Atinux** [comment]
Great PR @kevinmarrec :)

> Using nuxt-ts will always create the tsconfig.json where you're running the command, so it may cause unexpected behaviors like yarn nuxt-ts test/fixtures/typescript or yarn nuxt-ts examples/typescript creating a tsconfig.json in root folder.

About this, I guess we could add a prompt if no `tsconfig.json` is found to create one, WDTY?

**Comment by kevinmarrec** [comment]
@Atinux Seems to be a good solution, which library should be used for the prompt ? Have you any link to some code already using a prompt within Nuxt core (similar case where prompt is needed maybe ?) ?

**Comment by kevinmarrec** [comment]
And how can the prompt could be skipped with a `true` value in tests ?
I think we can have a simple test which check the creation and the content of the `tsconfig.json`, but I don't know yet how to achieve that

File: distributions/nuxt-ts/package.js [hunk 1]
```diff
@@ -18,6 +20,17 @@ export default {
       ])
 
       await pkg.writePackage()
+
+      if (pkg.options.suffix && pkg.options.linkedDependencies) {
+        const tsconfig = readJSONSync(pkg.resolvePath('tsconfig.json'))
```
Comment (pi0) [inline distributions/nuxt-ts/package.js]: Prefer async/await (non-Sync) methods when possible

Comment (kevinmarrec) [reply distributions/nuxt-ts/package.js]: Oops 🙈, gonna change that thanks !

**Comment by kevinmarrec** [comment]
This PR is ready for final review, and please check `Notes about future improvement` section of the PR description.

File: distributions/nuxt-ts/src/index.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    const { confirmGeneration } = await prompt({
+      type: 'confirm',
+      name: 'confirmGeneration',
+      message: `Missing ${chalk.bold.blue('tsconfig.json')} in ${rootDir === process.cwd() ? 'current directory' : chalk.bold.green(resolve(rootDir))}, generate it ?`,
+      initial: true,
+      skip: process.env.NODE_ENV === 'test'
+    })
+
+    if (confirmGeneration) {
+      await writeJSON(tsConfigPath, {
```
Comment (pi0) [inline distributions/nuxt-ts/src/index.js]: Idea: We can use `.json5` for tsconfig and annotate it with some comments. Like links to docs or indicating it is an autogenerated file.

File: distributions/nuxt-ts/bin/nuxt-ts.js [hunk 3]
```diff
@@ -3,16 +3,15 @@
...
+// rootDir should be set by a CLI helper to handle cases like `nuxt-ts path/to/project`
+const rootDir = process.cwd()
 
-const suffix = require('../package.json').name.includes('-edge') ? '-edge' : ''
-require('@nuxt/cli' + suffix).run()
-  .catch((error) => {
-    require('consola').fatal(error)
-    process.exit(2)
-  })
+require('..').generateTsConfigIfMissing(rootDir).then(() => {
```
[the remark this answers is not in the document]
Comment (kevinmarrec) [reply distributions/nuxt-ts/bin/nuxt-ts.js]: Yes thanks

File: distributions/nuxt-ts/src/index.js (hunk 2 above)
Comment (pi0) [reply distributions/nuxt-ts/src/index.js]: (Not sure. But probably typescript supports json5 by default for `.json` files too)

**Review by pi0 — DISMISSED** [review]
Overall LGTM. Left minor comments.

File: distributions/nuxt-ts/src/index.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+export async function generateTsConfigIfMissing(rootDir) {
+  const tsConfigPath = resolve(rootDir, 'tsconfig.json')
+
+  if (!existsSync(tsConfigPath)) {
+    const { confirmGeneration } = await prompt({
+      type: 'confirm',
+      name: 'confirmGeneration',
+      message: `Missing ${chalk.bold.blue('tsconfig.json')} in ${rootDir === process.cwd() ? 'current directory' : chalk.bold.green(resolve(rootDir))}, generate it ?`,
+      initial: true,
+      skip: process.env.NODE_ENV === 'test'
```
[the remark this answers is not in the document]
Comment (kevinmarrec) [reply distributions/nuxt-ts/src/index.js]: @pi0 Do you have a code suggestion to implement please ?

File: distributions/nuxt-ts/src/index.js (hunk 2 above)
Comment (kevinmarrec) [reply distributions/nuxt-ts/src/index.js]: Gonna take a look, thanks

File: distributions/nuxt-ts/src/index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (pi0) [reply distributions/nuxt-ts/src/index.js]:
```js
import env from 'std-env'

skip: env.minimal
```

Minimal is defined here:

https://github.com/jsless/std-env/blob/master/index.js#L66

[the remark this answers is not in the document]
Comment (kevinmarrec) [reply distributions/nuxt-ts/src/index.js]: Thanks ♥

File: distributions/nuxt-ts/src/index.js (hunk 2 above)
Comment (kevinmarrec) [reply distributions/nuxt-ts/src/index.js]: @pi0 I think we can't do it, I tried things but got issues.
Gonna add a `consola.info` with a link to the extended `tsconfig.json`

Comment (pi0) [reply distributions/nuxt-ts/src/index.js]: Ok. Thanks for your time on it ☺️

**Comment by kevinmarrec** [comment]
My last commit will make the extended `tsconfig.json` openable from terminals which support it :)

File: distributions/nuxt-ts/tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,33 @@
+{
+  "compilerOptions": {
+    "target": "esnext",
+    "moduleResolution": "node",
+    "lib": [
+      "esnext",
+      "esnext.asynciterable",
+      "dom"
+    ],
+    "esModuleInterop": true,
```
Comment (husayt) [inline distributions/nuxt-ts/tsconfig.json]: @kevinmarrec might be worth to add `resolveJsonModule": true,` to allow frictionless import of json files
```suggestion
    "esModuleInterop": true,
    "resolveJsonModule": true,
```

**Comment by kevinmarrec** [comment]
New version proposal :
https://github.com/nuxt/nuxt.js/compare/dev...kevinmarrec:nuxt-typescript-package

**Review by pi0 — DISMISSED** [review]

**Review by husayt — APPROVED** [review]
This is cool. Thanks @kevinmarrec

**Pull request merged**